services-core/src/gas_price/gnosis_safe.rs

```diff
@@ -0,0 +1,173 @@
+//! Gnosis Safe gas station `GasPriceEstimating` implementation.
+//! Api documentation at https://safe-relay.gnosis.io/ .
+
+use super::{linear_interpolation, GasPriceEstimating};
+use crate::http::{HttpClient, HttpFactory, HttpLabel};
+use anyhow::Result;
+use ethcontract::U256;
+use isahc::http::uri::Uri;
+use pricegraph::num;
+use serde::Deserialize;
+use serde_with::rust::display_fromstr;
+use std::{convert::TryInto, time::Duration};
+
+/// The default uris at which the gas station api is available under.
+const DEFAULT_MAINNET_URI: &str = "https://safe-relay.gnosis.io/api/v1/gas-station/";
+const DEFAULT_RINKEBY_URI: &str = "https://safe-relay.rinkeby.gnosis.io/api/v1/gas-station/";
+
+pub fn api_url_from_network_id(network_id: &str) -> Option<&'static str> {
+    match network_id {
+        "1" => Some(DEFAULT_MAINNET_URI),
+        "4" => Some(DEFAULT_RINKEBY_URI),
+        _ => None,
+    }
+}
+
+/// Gas prices in wei retrieved from the gas station. This is a result from the
+/// API call.
+#[derive(Deserialize, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct GasPrices {
+    pub last_update: String,
+    #[serde(with = "display_fromstr")]
+    pub lowest: f64,
+    #[serde(with = "display_fromstr")]
+    pub safe_low: f64,
+    #[serde(with = "display_fromstr")]
+    pub standard: f64,
+    #[serde(with = "display_fromstr")]
+    pub fast: f64,
+    #[serde(with = "display_fromstr")]
+    pub fastest: f64,
+}
+
+// The gnosis safe gas station looks at the gas price of all transactions in the last 200 blocks.
+// The fast gas price is the price at the 75th percentile of gas prices and so on.
+const FAST_PERCENTILE: f64 = 0.75;
+const STANDARD_PERCENTILE: f64 = 0.5;
+const SAFE_LOW_PERCENTILE: f64 = 0.3;
+
+// For this module we need to estimate confirmation times. So we need some way of converting the
+// percentiles into time. The standard percentile is 50% which means transactions at this gas price
+// were included in 50% of the blocks. Thus for every block a transaction at this price has a 0.5
+// chance to be included. We treat this as a geometric distribution in which the expected time for
+// the event to happen (transaction is included in block) is 1/p. Blocks happen every 15 seconds so
+// this is how we get a time estimate.
+// In reality this estimation can be problematic when gas prices are changing quickly. When prices
+// rising, the prices calculated based on the last 200 lagg behind the real price. For this reason
+// we insert two extra points for the linear interpolation below in estimate_limits.
+// We do not make use of the lowest and fastest gas price because they are too strong outliers. The
+// lowest gas price is skewed by miners including their own transactions at 1 gwei. The highest gas
+// price can be 1000 times the fast gas price which is not reasonable.
+// I suspect this happens when people try to ensure that two of their own transactions are included
```

Comment:
Love the explanation! 
ubernit: the speculative last two sentences are not really needed (in my experience miners aren't smart enough to look ahead and include a low gas tx in order to unblock a high gas tx).